src/charm.py

```diff
@@ -51,6 +52,12 @@ def __init__(self, *args: typing.Any):
         except CharmRelationDataInvalidError as exc:
             raise RuntimeError("Invalid relation data received.") from exc
 
+        try:
+            precondition.check(self, self.state)
+        except precondition.ConditionCheckError as exc:
+            self.unit.status = ops.WaitingStatus(exc.msg)
+            return
+
         self.actions_observer = actions.Observer(self, self.state)
         self.agent_observer = agent.Observer(self, self.state)
         self.cos_observer = cos.Observer(self)
```

Comment:
Can you confirm that this [line](https://github.com/canonical/jenkins-k8s-operator/blob/3e4c253492ea1f8fa7c0a33767cbaff515cc5905/src/charm.py#L58) is missing in `charm.py`? I find it strange that it's not present in the diff :confused:

Reply:
Without it the hook won't run so I'm wondering why the tests are passing

Reply:
It is probably due to the merge going into `feat/rotate-password` rather than `main`, I branched off from `feat/rotate-password` because there are quite a few changes that will lead to merge conflicts. Let me merge the `feat/rotate-password` with main and then merge this with updated `feat/rotate-password` :D

Reply:
Nvm, it was removed during merge conflict fixes, adding them back! sorry!